text/0000-deprecate-array-prototype-extensions.md

```diff
@@ -0,0 +1,149 @@
+---
+Stage: Initial
+Start Date: 2022-4-01
+Release Date: Unreleased
+Release Versions:
+  ember-source: vX.Y.Z
+  ember-data: vX.Y.Z
+Relevant Team(s): Ember.js
+RFC PR: 
+---
+
+# Deprecate array prototype extensions
+
+## Summary
+
+This RFC proposes to deprecate array prototype extensions.
+
+## Motivation
+
+Ember historically extended the prototypes of native Javascript arrays to implement `Ember.Enumerable`, `Ember.MutableEnumerable`, `Ember.MutableArray`, `Ember.Array`. This added convenient methods and properties, and also made Ember arrays automatically participate in the Ember Classic reactivity system.
+
+Those convenient methods increase the likelihood of becoming potential roadblocks for future built-in language extensions, and make it confusing for users to onboard: is it specifically part of Ember, or Javascript? Also with Ember Octane, the new reactivity system, those classic observable-based methods are no longer needed.
+
+We had deprecated [Functions](https://github.com/emberjs/rfcs/blob/master/text/0272-deprecation-native-function-prototype-extensions.md) and [Strings](https://github.com/emberjs/rfcs/blob/master/text/0236-deprecation-ember-string.md) prototype extensions. Array is the last step. And internally we had already been preferring generic array methods over prototype extensions ([epic](https://github.com/emberjs/ember.js/issues/15501)).
+
+Continuing in that direction, we should consider recommending the usage of native array functions as opposed to convenient prototype extension functions, and the usage of new tracked properties or `TrackedArray` over classic reactivity methods.
+
+## Transition Path
+
+For convenient methods like `without`, `sortBy`, `uniqBy` etc., the replacement functionality already exists either through generic array functions or utility libraries like [lodash](https://lodash.com), [Ramda](https://ramdajs.com), etc.
+
+For helper functions participating in the Ember classic reactivity system like `pushObject`, `removeObject`, the replacement functionality also already exists in the form of immutable update style with tracked propreties like `@tracked someArray = []`, or through utilizing `TrackedArray` from `tracked-built-ins`.
+
+During transition, we still allow users to use `A` from `@ember/array`.
+
+We don't need to build anything new specifically, however, the bulk of the transition will be
+focused on deprecating the array prototype extensions.
+
+## How We Teach This
+
+An entry to the [Deprecation Guides](https://deprecations.emberjs.com/v4.x) will be added outlining the different recommended transition strategies.
```

Comment:
When we do deprecation RFCs we always include the actual deprecation text for each thing we are deprecating. This is annoying, but when we break that rule we end up with really unhelpful things like [the `template-compiler.registerPlugin` deprecation](https://deprecations.emberjs.com/v3.x#toc_template-compiler-registerPlugin) for Ember v3, which at the time I'm leaving the comment here reads:

> Using class based template compilation plugins is deprecated. Please update to the functional style.
> 
> (This deprecation guide needs more details. You can help out by editing [this file](https://github.com/ember-learn/deprecation-app/blob/main/content/ember/v3/template-compiler-registerPlugin.md) and making a PR!)

This is obviously unhelpful. However, you should feel free to recruit folks to help you write those messages as part of the RFC process!

Reply:
Got it. Shall we create a folder v5 and have something like `deprecate-array-prototype-extensions` and list all the deprecated stuff and suggested resolution. Similar to this format: https://deprecations.emberjs.com/v3.x/#toc_ember-built-in-components-legacy-attribute-arguments. 

Listing all the deprecated methods/properties is not hard. I wonder how many details we would like to put for the suggestion part. We can go broad like `suggest use built-in method like `.any` -> `.some` and it's up to engineers to decide how they are going to convert it. Or give each method a possible equivalence.

Reply:
Yeah, that's a good idea for how to approach this!

Reply:
Created a [deprecation guide PR] (https://github.com/smilland/deprecation-app/pull/1) and referenced in draft RFC.